### DISCUSSION TITLE
git submodules and jj file list

### DISCUSSION THREAD

**Discussion by vcunat** [jj-vcs/jj#4063] [General]
I know that submodules aren't supported, but right now `jj file list` will output a path that doesn't exist.  That seems problematic; for example the suggested replacement for `git grep` is not only cumbersome to write, but it also causes errors because of this.

Maybe the submodules should get rendered as empty directories?  (I haven't thought it through, though)

**Comment by yuja** [comment]
Makes sense. Perhaps, submodule path can be emitted to stderr as a warning.

**Comment by arxanas** [comment]
- IIUC, it's not guaranteed that the submodule path _doesn't_ exist on disk, and conceivably you would want to be able to learn about them?
  - I think materializing a submodule as an empty directory if it doesn't exist at all might be okay for many workflows, but not sure. I don't know if it would impede Git materializing the submodule at a later time.
- Does this mean that we ought to generalize the warning/stderr output to cases where the path might not exist on disk?
  - I think this could include ignored-but-tracked files?
- Does `jj file list` include files that are tracked but not present in the sparse checkout?
  - If so, maybe the usage recommendation is really that `jj file list` might always return unmaterialized-but-tracked files, so you always need to filter them?
  - Alternatively, it might mean that `jj file list` wants an additional flag to filter to materialized files only, when specifically referring to the working-copy revision.
  - Tangential: there might be a reason that you want to list files in another revision according to the current sparse checkout rules?
- I guess, generally-speaking, `jj file list -r <rev>` might produce files which aren't guaranteed to exist on disk currently, so why should `-r @` be different?

**Reply by yuja** [reply]
`jj file list` works on commit, so the paths included in the output are not (and should not be) guaranteed to exist on disk. That said, I think it makes sense to exclude submodules because submodules are kinda trees, and `jj file list` doesn't print tree paths.

FWIW, at some point, we might add fileset functions like `tracked()` to filter the output.
